File: src/main/java/howtoprogram/words/WordNode.java
package howtoprogram.words;

import java.util.LinkedList;
import java.util.List;

/**
 * Generates one-letter variants of a word.
 */
public class WordNode {

    private String word;

    public WordNode(String word) {
        this.word = word;
    }

public List<WordNode> variantsByOneLetter(WordChecker checker) {
    List<WordNode> result = new LinkedList<>();
    //For each position in the word...
    for (int i = 0; i < word.length(); i++) {
        //...add the variants made by changing
        //just the letter at that position.
        addVariantsAtPosition(i, result, checker);
    }
    return result;
}

private void addVariantsAtPosition(int position, List<WordNode> nodes, WordChecker checker) {
    String alphabet = "abcdefghijklmnopqrstuvwxyz";
    char original = word.charAt(position);
    for (int i = 0; i < alphabet.length(); i++) {
        char fromAlphabet = alphabet.charAt(i);
        if (original != fromAlphabet) {
            StringBuilder sb = new StringBuilder(word);
            sb.deleteCharAt(position);
            sb.insert(position, fromAlphabet);
            String toAdd = sb.toString();
            if (checker.isPreviouslyUnseenValidWord(toAdd)) {
                nodes.add(new WordNode(toAdd));
            }
        }
    }
}

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        WordNode wordNode = (WordNode) o;

        return word.equals(wordNode.word);

    }

    @Override
    public int hashCode() {
        return word.hashCode();
    }
}

File: src/main/kotlin/pfb/words/WordNode.kt
package pfb.words

/**
 * Generates one-letter variants of a word.
 */
data class WordNode(val text: String) {
    fun variantsByOneLetter(checker: WordChecker): List<WordNode> {
        val result = mutableListOf<WordNode>()
        //For each position in the word...
        for (i in 0..text.length - 1) {
            //...add the variants made by changing
            //just the letter at that position.
            addVariantsAtPosition(i, result, checker)
        }
        return result
    }

    private fun addVariantsAtPosition(position: Int, nodes: MutableList<WordNode>, checker: WordChecker) {
        val textBeforePosition = text.substring(0, position)
        val textAfterPosition = if (text.length > 1) text.substring(position + 1, text.length) else ""
        val originalLetter = text[position]
        for (fromAlphabet in "abcdefghijklmnopqrstuvwxyz") {
            if (originalLetter != fromAlphabet) {
                val variant = textBeforePosition + fromAlphabet + textAfterPosition
                if (checker.isPreviouslyUnseenValidWord(variant)) {
                    nodes.add(WordNode(variant))
                }
            }
        }
    }
}